dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php

```diff
@@ -0,0 +1,213 @@
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\Exception\StateException;
+use Magento\Framework\Registry;
+use Magento\ProductVariantDataExporter\Model\ProductVariantFeedInterface;
+use Magento\TestFramework\Helper\Bootstrap;
+
+/**
+ * Test class for Products variants message bus
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class ProductVariantsTest extends StorefrontTestsAbstract
+{
+    const CONFIGURABLE_SKU = 'configurable';
+    const SIMPLE1_SKU = 'simple_10';
+    const SIMPLE2_SKU = 'simple_20';
+
+    const STORE_CODE = 'default';
+    const ERROR_MESSAGE = 'No products variants for product with id %s are found in catalog.';
+
+    /**
+     * @var ProductVariantsConsumer
+     */
+    private $productVariantsConsumer;
+
+    /**
+     * @var VariantService
+     */
+    private $variantService;
+
+    /**
+     * @var ProductVariantRequestInterface
+     */
+    private $variantsGetRequestInterface;
+
+    /**
+     * @var ChangedEntitiesMessageBuilder
+     */
+    private $messageBuilder;
+
+    /**
+     * @var ProductVariantFeedInterface
+     */
+    private $productVariantFeed;
+
+    /**
+     * @var ProductRepositoryInterface
+     */
+    private $productRepository;
+
+    /**
+     * @var ProductVariantResponseArrayMapper
+     */
+    private $responseArrayMapper;
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->productVariantsConsumer = Bootstrap::getObjectManager()->create(ProductVariantsConsumer::class);
+        $this->variantService = Bootstrap::getObjectManager()->create(VariantService::class);
+        $this->variantsGetRequestInterface = Bootstrap::getObjectManager()->create(
+            ProductVariantRequestInterface::class
+        );
+        $this->messageBuilder = Bootstrap::getObjectManager()->create(ChangedEntitiesMessageBuilder::class);
+        $this->productVariantFeed = Bootstrap::getObjectManager()->get(FeedPool::class)->getFeed('variants');
+        $this->productRepository = Bootstrap::getObjectManager()->create(ProductRepositoryInterface::class);
+        $this->responseArrayMapper = Bootstrap::getObjectManager()->create(ProductVariantResponseArrayMapper::class);
+    }
+
+    /**
+     * Validate save and delete product variant operations
+     *
+     * @magentoApiDataFixture Magento/ConfigurableProduct/_files/product_configurable_sku.php
+     * @magentoDbIsolation disabled
+     * @throws NoSuchEntityException
+     * @throws StateException
+     * @throws \Throwable
+     */
+    public function testSaveAndDeleteProductVariant(): void
+    {
+        $configurable = $this->getProduct(self::CONFIGURABLE_SKU);
+        $configurableId = $configurable->getId();
+        $simple1 = $this->getProduct(self::SIMPLE1_SKU);
+        $simple1Id = $simple1->getId();
+        $simple2 = $this->getProduct(self::SIMPLE2_SKU);
+        $simple2Id = $simple2->getId();
```

Comment:
add more variants to configurable product

Reply:
Done